File: frontend/src/lib/outlook-api.ts
import { apiClient, ApiHttpError } from "@/lib/api-client";

export type OutlookSource = {
  title: string;
  url: string;
  publisher: string;
  published_at?: string | null;
};

export type OutlookInference = {
  volatility_view: string;
  movement_scenarios: string[];
  confidence: "low" | "medium" | "high";
  caveats: string[];
};

export type OutlookStrategyIdea = {
  tag: string;
  rationale: string;
  risk_note: string;
};

export type OutlookSummaryCategory = "macro_global" | "macro_local" | "positioning";

export type OutlookSummaryItem = {
  category: OutlookSummaryCategory;
  text: string;
};

export type OutlookWarning = {
  error_code: string;
  message: string;
  stale_response_served: boolean;
  upstream_status?: number | null;
};

export type OutlookResponse = {
  outlook_type: "market";
  as_of: string;
  english_only: boolean;
  disclaimer: string;
  summary: OutlookSummaryItem[];
  inference: OutlookInference;
  strategy_ideas: OutlookStrategyIdea[];
  sources: OutlookSource[];
  warning?: OutlookWarning;
};

/**
 * The portal's summary schema was previously a flat string[]. Deployments upgrade
 * independently of the portal, so this repo's frontend may briefly receive either
 * shape from a not-yet-upgraded portal cache -- normalize defensively rather than
 * assume both sides are always in lockstep.
 */
function normalizeOutlookSummary(raw: unknown): OutlookSummaryItem[] {
  if (!Array.isArray(raw)) return [];
  return raw.map((item) =>
    typeof item === "string" ? { category: "macro_global" as const, text: item } : (item as OutlookSummaryItem),
  );
}

export function getMarketOutlook(forceRefresh = false) {
  const suffix = forceRefresh ? "?force_refresh=true" : "";
  return apiClient.get<OutlookResponse>(`/api/outlook/market${suffix}`).then((data) => ({
    ...data,
    summary: normalizeOutlookSummary(data.summary),
  }));
}

/** User-visible copy for outlook fetch failures. */
export function outlookFetchErrorMessage(err: unknown): string {
  if (!(err instanceof ApiHttpError)) {
    return err instanceof Error ? err.message : "Could not load market outlook.";
  }
  return err.message;
}

File: backend/src/icici_breeze_backend/app/domain/outlook_api.py
"""API schemas for GenAI market outlook."""
from __future__ import annotations

from typing import Literal, Optional

from pydantic import BaseModel, Field


OUTLOOK_DISCLAIMER = (
    "AI-generated outlook. Informational only. Not investment advice. Verify with primary sources."
)


class OutlookSource(BaseModel):
    title: str
    url: str
    publisher: str
    published_at: Optional[str] = None


class OutlookInference(BaseModel):
    volatility_view: str
    movement_scenarios: list[str] = Field(default_factory=list)
    confidence: Literal["low", "medium", "high"]
    caveats: list[str] = Field(default_factory=list)


class OutlookStrategyIdea(BaseModel):
    tag: str
    rationale: str
    risk_note: str


class OutlookResponse(BaseModel):
    outlook_type: Literal["market"]
    as_of: str
    english_only: bool = True
    disclaimer: str = OUTLOOK_DISCLAIMER
    summary: list[str] = Field(default_factory=list)
    inference: OutlookInference
    strategy_ideas: list[OutlookStrategyIdea] = Field(default_factory=list)
    sources: list[OutlookSource] = Field(default_factory=list)


class OutlookErrorResponse(BaseModel):
    error_code: str
    message: str
    stale_response_served: bool = False
